### ISSUE TITLE
valgrind shows memory leaks inside SDK

### ISSUE THREAD

**Issue by shekharhimanshu** [aws/aws-iot-device-sdk-cpp#23]
Hello,

I ran the [PubSub sample](https://github.com/aws/aws-iot-device-sdk-cpp/tree/master/samples/PubSub) against valgrind and it looks like the SDK leaks memory (besides other issues like Mismatched free() etc). I am attaching the valgrind log for reference.

Please look into it.
[log_memory_leak.txt](https://github.com/aws/aws-iot-device-sdk-cpp/files/1097414/log_memory_leak.txt)

**Comment by chaurah** [comment]
Hi @shekharhimanshu, 
Thanks for this update. We did look into these and there wasn't any major red flag in our analysis at the time. The issues were mostly localized to the mbedtls/openssl implementations. I will however go through these again with valgrind and will also run it on all our sample code to ensure something hasn't been overlooked. If you have any specific locations where you are already sure of mismatched free/memory leaks please do let me know and I will add those tasks as high priority for a bugfix release.

Rahul

**Comment by shekharhimanshu** [comment]
>We did look into these and there wasn't any major red flag in our analysis at the time. The issues were mostly localized to the mbedtls/openssl implementations.

You are spot on about leaks inside mbedtls implementation which IMHO should be fixed as well, but there are seems to be other leaks as well. For example, it looks like the network connection instance is not getting freed. See the log snippet below:

```
==15665== 
==15665== 3,592 bytes in 1 blocks are indirectly lost in loss record 101 of 105
==15665==    at 0x4C2E0EF: operator new(unsigned long) (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==15665==    by 0x53C926: __gnu_cxx::new_allocator<std::_Sp_counted_ptr_inplace<awsiotsdk::network::MbedTLSConnection, std::allocator<awsiotsdk::network::MbedTLSConnection>, (__gnu_cxx::_Lock_policy)2> >::allocate(unsigned long, void const*) (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x53C472: std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<awsiotsdk::network::MbedTLSConnection, std::allocator<awsiotsdk::network::MbedTLSConnection>, (__gnu_cxx::_Lock_policy)2> > >::allocate(std::allocator<std::_Sp_counted_ptr_inplace<awsiotsdk::network::MbedTLSConnection, std::allocator<awsiotsdk::network::MbedTLSConnection>, (__gnu_cxx::_Lock_policy)2> >&, unsigned long) (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x53BD4D: std::__allocated_ptr<std::allocator<std::_Sp_counted_ptr_inplace<awsiotsdk::network::MbedTLSConnection, std::allocator<awsiotsdk::network::MbedTLSConnection>, (__gnu_cxx::_Lock_policy)2> > > std::__allocate_guarded<std::allocator<std::_Sp_counted_ptr_inplace<awsiotsdk::network::MbedTLSConnection, std::allocator<awsiotsdk::network::MbedTLSConnection>, (__gnu_cxx::_Lock_policy)2> > >(std::allocator<std::_Sp_counted_ptr_inplace<awsiotsdk::network::MbedTLSConnection, std::allocator<awsiotsdk::network::MbedTLSConnection>, (__gnu_cxx::_Lock_policy)2> >&) (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x53B5DF: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<awsiotsdk::network::MbedTLSConnection, std::allocator<awsiotsdk::network::MbedTLSConnection>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool>(std::_Sp_make_shared_tag, awsiotsdk::network::MbedTLSConnection*, std::allocator<awsiotsdk::network::MbedTLSConnection> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool&&) (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x53A8C1: std::__shared_ptr<awsiotsdk::network::MbedTLSConnection, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<awsiotsdk::network::MbedTLSConnection>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool>(std::_Sp_make_shared_tag, std::allocator<awsiotsdk::network::MbedTLSConnection> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool&&) (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x539748: std::shared_ptr<awsiotsdk::network::MbedTLSConnection>::shared_ptr<std::allocator<awsiotsdk::network::MbedTLSConnection>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool>(std::_Sp_make_shared_tag, std::allocator<awsiotsdk::network::MbedTLSConnection> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool&&) (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x53822A: std::shared_ptr<awsiotsdk::network::MbedTLSConnection> std::allocate_shared<awsiotsdk::network::MbedTLSConnection, std::allocator<awsiotsdk::network::MbedTLSConnection>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool>(std::allocator<awsiotsdk::network::MbedTLSConnection> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool&&) (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x5367C1: std::shared_ptr<awsiotsdk::network::MbedTLSConnection> std::make_shared<awsiotsdk::network::MbedTLSConnection, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, std::chrono::duration<long, std::ratio<1l, 1000l> >&, bool&&) (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x533AB9: awsiotsdk::samples::PubSub::InitializeTLS() (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x533C14: awsiotsdk::samples::PubSub::RunSample() (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
==15665==    by 0x5346BA: main (in /home/quiver/Develop/third-party/aws-iot-device-sdk-cpp-1.1.0/cmake_build/bin/pub-sub-sample)
```

>I will however go through these again with valgrind and will also run it on all our sample code to ensure something hasn't been overlooked.

Please do.

>If you have any specific locations where you are already sure of mismatched free/memory leaks please do let me know and I will add those tasks as high priority for a bugfix release.

I don't have any other specific locations besides the ones already present in the log file. Will look for possible leak cause myself as well.

**Comment by shekharhimanshu** [comment]
>You are spot on about leaks inside mbedtls implementation which IMHO should be fixed as well, but there are seems to be other leaks as well. For example, it looks like the network connection instance is not getting freed.

I found the code causing this leak and have submitted a fix.

**Comment by vareddy** [comment]
Hi @shekharhimanshu,
Your pull request for the memory leak has been included in v1.1.1.
Thanks!
Varun
